kafka_checked/src/main.rs

```diff
@@ -0,0 +1,363 @@
+                            xml: xmlmsg
+                        }
+                    );
+                    break 'message;
+                }
+                let kvalobs_xmlmsg = &xmlmsg[(loc_end_xml_tag.unwrap() + 2)..(xmlmsg.len())];
+
+                let item: KvalobsData = from_str(kvalobs_xmlmsg).unwrap();
+
+                // get the useful stuff out of this struct
+                for station in item.station {
+                    for typeid in station.typeid {
+                        'obstime: for obstime in typeid.obstime {
+                            match NaiveDateTime::parse_from_str(&obstime.val, "%Y-%m-%d %H:%M:%S") {
+                                Ok(obs_time) => {
+                                    for tbtime in obstime.tbtime {
+                                        // NOTE: this is "table time" which can vary from the actual observation time,
+                                        // its the first time in entered the db in kvalobs
+                                        let tb_time = NaiveDateTime::parse_from_str(
+                                            &tbtime.val,
+                                            "%Y-%m-%d %H:%M:%S%.6f",
+                                        );
+                                        if tb_time.is_err() {
+                                            eprintln!(
+                                                "{}",
+                                                Error::IssueParsingTime(
+                                                    tb_time.expect_err("cannot parse tbtime")
+                                                )
+                                            );
+                                            // currently not using this, so can just log and keep going?
+                                        }
+                                        /*
+                                        // TODO: Do we want to handle text data at all, it doesn't seem to be QCed
+                                        if let Some(textdata) = tbtime.kvtextdata {
+                                            println!(
+                                                "station {:?}, typeid {:?}, textdata: {:?}",
+                                                station.val, typeid.val, textdata
+                                            );
+                                        }
+                                        */
+                                        for sensor in tbtime.sensor {
+                                            for level in sensor.level {
+                                                if let Some(data) = level.kvdata {
+                                                    for d in data {
+                                                        // but change the sensor and level back to null if they are 0 
+                                                        // 0 is the default for kvalobs, but through obsinn its actually just missing
```

Comment:
Hope this is a sensible way to convert 0 back to a None. The other thing we could do is convert paramid to paramcode so we could look at the obsinn label...?

Reply:
Looks good